### PR TITLE
Change compareStr test to allow for length check (line 890)

### PR DESCRIPTION
**Pull request by TwoFistedJustice**
Sorry for not doing this earlier. I was prepping for Hack Reactor and didn't see your request for change until today.

### CODE REVIEW

File: spec/part1.js [hunk 1]
```diff
@@ -887,7 +887,7 @@
       });
 
       it('should use recursion by calling self', function() {
-        compareStr('house', 'houses');
+        compareStr('houses', 'houses');
```
Comment (mybrainishuge) [inline spec/part1.js]: If I recall correctly, you suggested this change on the assumption that people will implement this function with the "optimization" of including a direct string comparison. A direct string comparison goes both ways. You can check for equality or inequality. We can't assume which approach users will take. There's no way around using the equality operator when implementing this function, but using `wholeString === wholeString` sidesteps the purpose of this exercise which is to practice recursion.

Or am I remembering your argument incorrectly?

File: src/recursion.js [hunk 2]
```diff
@@ -71,7 +71,7 @@ var multiply = function(x, y) {
 };
 
 // 13. Write a function that divides two numbers without using the / operator or
-// Math methods.
+// Math methods to arrive at an approximate quotient (ignore decimal endings).
```
Comment (mybrainishuge) [inline src/recursion.js]: This change is already merged into `master`. I'm a little confused why it's appearing in this PR. Have you rebased with `master`?

File: spec/part1.js (hunk 1 above)
Comment (TwoFistedJustice) [reply spec/part1.js]: That is not correct. 

I suggest it to allow for length comparison. Since there is no specification to ignore spaces a simple length comparison would be a valid optimization.

Here is my solution, which fails the original recursion count test due to the length test

`var compareStr = function(str1, str2) {

// Set a length variable and check that lengths are equal - set to a falsy value if they are not
  var len = str1.length === str2.length
    ? str1.length
    : null;
  
// set the base case using length
  if (len === 0) {
    return true;
  } else {
  // check that both ends are equal, recurse if they are equal
    return (len !== null && str1[0] === str2[0] && str1[len - 1] && str2[len - 1])
      ? compareStr(str1.slice(1, len - 2), str2.slice(1, len - 2))
      : false;    
  }
};`

File: src/recursion.js (hunk 2 above)
Comment (TwoFistedJustice) [reply src/recursion.js]: Since I didn't realize that was even a thing, I'm going to go with "no".

Will do going forward.

**Comment by TwoFistedJustice** [comment]
I am going to close this PR and then try to correct my commit history, rebase with master, etc. 

I will re-create the PR after I have it straightened out.

File: spec/part1.js (hunk 1 above)
Comment (mybrainishuge) [reply spec/part1.js]: Still, I'm not sure how I feel about accounting for this optimization. Checking for overall length isn't that far off from checking the equality of the overall string. This repo is intended to encourage people to think about the recursive stack. Finding ways to short-circuit diving into the recursive stack sidesteps that purpose and redirects the focus on optimization.

Let me give you another example. Most people find recursion much easier to think about if the function accepts an additional argument where you pass down the "result" and mutate it on each recursive call. I wanted to force people to think about the concept of how to build a result using the return value of a recursive call which most people find more difficult to grasp. This is why you see these checking for the number of arguments used.

You're completely right that optimizations are good practice for recursion and it's great that you're thinking about it. However, defining base cases is not one of the more complex aspects of recursion. Picturing the stack and passing values up/down it is more difficult for most people to wrap their head around. I want to push people into that mindset because the more you practice it, the more comfortable it becomes.

Comment (TwoFistedJustice) [reply spec/part1.js]: Well that makes sense.

One thing I noticed is that as it is, one MUST check from the beginning of the string. So if someone checks the last char first, they will fail the test for recursion. -- "e" vs "s"

May I suggest changing this test so that the string args are the same on either end, but different in the middle?

Something like "abba" and "ababa"

This should pass the recursion test regardless of which end they start at; or even if they check both ends simultaneously.

Comment (mybrainishuge) [reply spec/part1.js]: That's true, but try not to get lost in identifying edge cases. The tests would be more robust, but I don't expect many people will start from the end of the string. Also, keep in mind that the tests are visible to everyone. You're an excellent example of this. Your code wasn't passing, so what did you do? You dug deeper by looking at the tests. That's a sign of a good engineer.

If you still want to open a PR to adjust this test for this latest suggestion, go ahead.
